File: solutions/cpp/Tail.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <fstream>
#include <iostream>
#include <string>

using std::cout;
using std::endl;
using std::fstream;
using std::ios;
using std::ofstream;
using std::string;
using std::to_string;

// @include
string tail(const string& file_name, int N) {
  fstream file_ptr(file_name.c_str());

  file_ptr.seekg(0, ios::end);
  int file_size = file_ptr.tellg(), newline_count = 0;
  string last_N_lines;
  // Reads file in reverse looking for '\n'.
  for (int i = 0; i < file_size; ++i) {
    file_ptr.seekg(-1 - i, ios::end);
    char c;
    file_ptr.get(c);
    if (c == '\n') {
      ++newline_count;
      if (newline_count >= N) {
        break;
      }
    }
    last_N_lines.push_back(c);
  }

  reverse(last_N_lines.begin(), last_N_lines.end());
  return last_N_lines;
}
// @exclude

void SmallTest() {
  ofstream ofs("test_Tail.txt");
  string L1 = "The first line";
  string L2 = "The second line";
  string L3 = "The third line";
  string L4 = "The fourth line";
  ofs << L1 << endl;
  ofs << L2 << endl;
  ofs << L3 << endl;
  ofs << L4;
  ofs.close();
  auto res = tail("test_Tail.txt", 1);
  assert(!res.compare(L4));
  res = tail("test_Tail.txt", 2);
  assert(!res.compare(L3 + "\n" + L4));
  res = tail("test_Tail.txt", 3);
  assert(!res.compare(L2 + "\n" + L3 + "\n" + L4));
  res = tail("test_Tail.txt", 4);
  assert(!res.compare(L1 + "\n" + L2 + "\n" + L3 + "\n" + L4));
  remove("test_Tail.txt");
}

int main(int argc, char* argv[]) {
  SmallTest();
  cout << "Usage: file name and tail count" << endl;
  int tail_count = 10;
  string file_name;
  if (argc == 2) {
    file_name = argv[1];
  } else if (argc == 3) {
    file_name = argv[1];
    tail_count = atoi(argv[2]);
  } else {
    exit(1);
  }
  string output = tail(file_name, tail_count);
  cout << output;
  string command = "tail -n " + to_string(tail_count) + " " + file_name;
  cout << "command = " << command << endl;
  system(command.c_str());
  return 0;
}

File: static/python/tail.py
# Tail.cpp 848813e190b1b85a8e75107fe8513c3be38ad1a9
import sys
import os


# @include
def tail(file_name, N):
    with open(file_name, 'rb') as file_ptr:
        # Note: you can seek from the file's end only in binary mode.
        file_size = file_ptr.seek(0, os.SEEK_END)
        newline_count = 0
        last_n_lines = []
        # Reads file in reverse looking for '\n'.
        for i in range(file_size):
            file_ptr.seek(-1 - i, os.SEEK_END)
            c = file_ptr.read(1)
            if c == b'\n':
                newline_count += 1
                if newline_count > N:
                    break
            last_n_lines.append(c)

    return b''.join(reversed(last_n_lines))


# @exclude


def small_test():
    with open('test_Tail.txt', 'wb') as ofs:
        ofs.write(
            b"A free soft copy sample of Elements of Programming Interviews can be found at bit.ly/epi-sample\n\n"
            b"The core of Elements of Programming Interviews (EPI) is a collection of 300 problems with detailed solutions, including over 150 figures and 250 tested programs. The problems are representative of the questions asked at interviews at the most exciting companies. They are well-motivated, thought-provoking and fun to solve!\n\n"
            b"EPI includes a summary of the nontechnical aspects of interviewing, including common mistakes, strategies for a great interview, the interviewer's perspective, negotiating the best offer, and much more.\n\n"
            b"Since different candidates have different time constraints, EPI includes a study guide with several study scenarios, ranging from weekend Hackathon to semester long preparation with a recommended a subset of problems for each scenario.\n\n")
    res = tail('test_Tail.txt', 3)
    assert res == b'\nSince different candidates have different time constraints, EPI includes a study guide with several study scenarios, ranging from weekend Hackathon to semester long preparation with a recommended a subset of problems for each scenario.\n\n'
    os.remove('test_Tail.txt')


def main():
    small_test()
    print('Usage: file name and tail count')
    if len(sys.argv) == 2:
        file_name = sys.argv[1]
        tail_count = 10
    elif len(sys.argv) == 3:
        file_name = sys.argv[1]
        tail_count = int(sys.argv[2])
    else:
        sys.exit(1)

    output = tail(file_name, tail_count)
    print(output)
    command = 'tail -n %d %s' % (tail_count, file_name)
    print('command =', command)
    os.system(command)


if __name__ == '__main__':
    main()
